### PR TITLE
3888 typed context and parser

### PR DESCRIPTION
**Pull request by jberthold**
* Introduces a JSON-enabled type for the log context
* Uses the new type in all  booster logging
* Modifies `count-aborts` to parse the json data using the new type. Performance is on par after switching to TH-derived `FromJSON` instances.

Conversion of `kore-rpc` logging is left for future work.
Part of  #3888

### CODE REVIEW

File: booster/library/Booster/Log.hs [hunk 1]
```diff
@@ -100,78 +138,84 @@ logMessage' a =
...
-    m
+withTermContext t@(Term attrs _) m =
+    withContext_ (CLWithId . CtxTerm . ShortId $ showHashHex attrs.hash) $ do
+        withContext CtxKoreTerm $ logMessage t
+        m
 
 withPatternContext :: LoggerMIO m => Pattern -> m a -> m a
 withPatternContext Pattern{term, constraints} m =
-    let t' = foldl' AndTerm term $ Set.toList $ Set.map coerce constraints
+    let t' = foldl' AndTerm term $ Set.toList $ Set.map coerce constraints -- FIXME
```
Comment (jberthold) [inline booster/library/Booster/Log.hs:158]: I noticed this earlier while trying to copy/paste a `KoreJson` pattern from a log into a file for custom requests.
* The output is not correctly sorted (we would need to introduce `#Equals{SortBool, <sort-of-term>}(true, _)` around the predicates)
* That would  mean we cannot use `AndTerm` any more
We should probably go to `KoreJson` first  (externalise the pattern) for the output, but then we need to come up with a stable hash for logging purposes

Comment (geo2a) [reply booster/library/Booster/Log.hs:158]: Does it cause issues when internalising? I.e. when you copy/pasted `KoreJson` from a log line and tried to send that to the server it reported an error?

File: booster/library/Booster/SMT/Interface.hs [hunk 2]
```diff
@@ -168,13 +168,13 @@ getModelFor ::
     Map Variable Term -> -- supplied substitution
     io (Either SMTError (Either SMT.Response (Map Variable Term)))
 getModelFor ctxt ps subst
-    | null ps && Map.null subst = Log.withContext "smt" $ do
-        Log.logMessage ("No constraints or substitutions to check, returning Sat" :: Text)
+    | null ps && Map.null subst = Log.withContext Log.CtxSMT $ do
+        Log.logMessage ("No Constraints Or Substitutions To Check, Returning Sat" :: Text)
```
Comment (geo2a) [inline booster/library/Booster/SMT/Interface.hs:172]: Such Capitalism!

File: booster/library/Booster/Log.hs (hunk 1 above)
Comment (jberthold) [reply booster/library/Booster/Log.hs:158]: The server rejects the predicates having a wrong sort IIRC, but I would have to re-try this to be sure.

File: dev-tools/kore-rpc-dev/Server.hs [hunk 3]
```diff
@@ -230,7 +230,7 @@ main = do
                 kore@KoreServer{runSMT} <-
                     mkKoreServer Log.LoggerEnv{logAction} clOPts koreSolverOptions
                 runBoosterLogger $
-                    Booster.Log.withContext "proxy" $
+                    Booster.Log.withContext CtxKore $
```
Comment (geo2a) [inline dev-tools/kore-rpc-dev/Server.hs:233]: :+1:

**Review by geo2a — APPROVED** [review]
Monumental work and a major improvement!

File: booster/library/Booster/SMT/Interface.hs (hunk 2 above)
Comment (jberthold) [reply booster/library/Booster/SMT/Interface.hs:172]: :rofl: Emacs shortcuts going rogue.... I remember pressing the wrong keys (`M^c`) here... corrected everything that the compiler cares about but not the string contents :smile:

Comment (jberthold) [reply booster/library/Booster/SMT/Interface.hs:172]:
```suggestion
        Log.logMessage ("No constraints or substitutions to check, returning Sat" :: Text)
```

**Review by goodlyrottenapple — APPROVED** [review]

**Pull request merged**